File: app/src/main/java/com/example/aichat/SessionChatOptionsStore.java
package com.example.aichat;

import android.content.Context;

public class SessionChatOptionsStore {

    private final SessionChatOptionsDao dao;

    public SessionChatOptionsStore(Context context) {
        dao = AppDatabase.getInstance(context).sessionChatOptionsDao();
    }

    public SessionChatOptions get(String sessionId) {
        if (sessionId == null || sessionId.isEmpty()) return defaultOptions();
        SessionChatOptionsEntity entity = dao.get(sessionId);
        SessionChatOptions out = entity != null ? fromEntity(entity) : new SessionChatOptions();
        out.streamOutput = true; // always force true per existing contract
        return out;
    }

    public boolean has(String sessionId) {
        if (sessionId == null || sessionId.isEmpty()) return false;
        return dao.has(sessionId) > 0;
    }

    public void save(String sessionId, SessionChatOptions options) {
        if (sessionId == null || sessionId.isEmpty() || options == null) return;
        options.streamOutput = true;
        dao.upsert(toEntity(sessionId, options));
    }

    public void remove(String sessionId) {
        if (sessionId == null || sessionId.isEmpty()) return;
        dao.delete(sessionId);
    }

    // --- Conversion helpers ---

    private static SessionChatOptions defaultOptions() {
        SessionChatOptions opts = new SessionChatOptions();
        opts.streamOutput = true;
        return opts;
    }

    private static SessionChatOptionsEntity toEntity(String sessionId, SessionChatOptions opts) {
        SessionChatOptionsEntity entity = new SessionChatOptionsEntity();
        entity.sessionId = sessionId;
        entity.sessionTitle = opts.sessionTitle != null ? opts.sessionTitle : "";
        entity.sessionAvatar = opts.sessionAvatar != null ? opts.sessionAvatar : "";
        entity.modelKey = opts.modelKey != null ? opts.modelKey : "";
        entity.systemPrompt = opts.systemPrompt != null ? opts.systemPrompt : "";
        entity.stop = opts.stop != null ? opts.stop : "";
        entity.contextMessageCount = opts.contextMessageCount;
        entity.googleThinkingBudget = opts.googleThinkingBudget;
        entity.temperature = opts.temperature;
        entity.topP = opts.topP;
        entity.streamOutput = true;
        entity.autoChapterPlan = opts.autoChapterPlan;
        entity.thinking = opts.thinking;
        return entity;
    }

    private static SessionChatOptions fromEntity(SessionChatOptionsEntity entity) {
        SessionChatOptions opts = new SessionChatOptions();
        opts.sessionTitle = entity.sessionTitle != null ? entity.sessionTitle : "";
        opts.sessionAvatar = entity.sessionAvatar != null ? entity.sessionAvatar : "";
        opts.modelKey = entity.modelKey != null ? entity.modelKey : "";
        opts.systemPrompt = entity.systemPrompt != null ? entity.systemPrompt : "";
        opts.stop = entity.stop != null ? entity.stop : "";
        opts.contextMessageCount = entity.contextMessageCount;
        opts.googleThinkingBudget = entity.googleThinkingBudget;
        opts.temperature = entity.temperature;
        opts.topP = entity.topP;
        opts.streamOutput = true;
        opts.autoChapterPlan = entity.autoChapterPlan;
        opts.thinking = entity.thinking;
        return opts;
    }
}

File: app/src/main/java/com/example/aichat/SessionChatOptionsStore.kt
package com.example.aichat

import android.content.Context

class SessionChatOptionsStore(context: Context) {

    private val dao: SessionChatOptionsDao = AppDatabase.getInstance(context).sessionChatOptionsDao()

    fun get(sessionId: String?): SessionChatOptions {
        if (sessionId.isNullOrEmpty()) return defaultOptions()
        val entity = dao.get(sessionId)
        val out = if (entity != null) fromEntity(entity) else SessionChatOptions()
        out.streamOutput = true // always force true per existing contract
        return out
    }

    fun has(sessionId: String?): Boolean {
        if (sessionId.isNullOrEmpty()) return false
        return dao.has(sessionId) > 0
    }

    fun save(sessionId: String?, options: SessionChatOptions?) {
        if (sessionId.isNullOrEmpty() || options == null) return
        options.streamOutput = true
        dao.upsert(toEntity(sessionId, options))
    }

    fun remove(sessionId: String?) {
        if (sessionId.isNullOrEmpty()) return
        dao.delete(sessionId)
    }

    // --- Conversion helpers ---

    companion object {
        private fun defaultOptions(): SessionChatOptions {
            val opts = SessionChatOptions()
            opts.streamOutput = true
            return opts
        }

        private fun toEntity(sessionId: String, opts: SessionChatOptions): SessionChatOptionsEntity {
            val entity = SessionChatOptionsEntity()
            entity.sessionId = sessionId
            entity.sessionTitle = opts.sessionTitle ?: ""
            entity.sessionAvatar = opts.sessionAvatar ?: ""
            entity.sessionAvatarImageBase64 = opts.sessionAvatarImageBase64
            entity.modelKey = opts.modelKey ?: ""
            entity.systemPrompt = opts.systemPrompt ?: ""
            entity.stop = opts.stop ?: ""
            entity.contextMessageCount = opts.contextMessageCount
            entity.googleThinkingBudget = opts.googleThinkingBudget
            entity.temperature = opts.temperature
            entity.topP = opts.topP
            entity.maxTokens = opts.maxTokens
            entity.frequencyPenalty = opts.frequencyPenalty
            entity.presencePenalty = opts.presencePenalty
            entity.topK = opts.topK
            entity.streamOutput = true
            entity.autoChapterPlan = opts.autoChapterPlan
            entity.thinking = opts.thinking
            entity.autoChatEnabled = opts.autoChatEnabled
            entity.proactiveCountToday = opts.proactiveCountToday
            entity.proactiveResetDate = opts.proactiveResetDate
            entity.proactiveDailyBudget = opts.proactiveDailyBudget
            entity.inkosEnabled = opts.inkosEnabled
            entity.inkosBookId = opts.inkosBookId
            entity.inkosSubtype = opts.inkosSubtype
            entity.inkosBookRulesYaml = opts.inkosBookRulesYaml
            entity.inkosTargetChapters = opts.inkosTargetChapters
            entity.inkosChapterWordCount = opts.inkosChapterWordCount
            return entity
        }

        private fun fromEntity(entity: SessionChatOptionsEntity): SessionChatOptions {
            val opts = SessionChatOptions()
            opts.sessionTitle = entity.sessionTitle ?: ""
            opts.sessionAvatar = entity.sessionAvatar ?: ""
            opts.sessionAvatarImageBase64 = entity.sessionAvatarImageBase64
            opts.modelKey = entity.modelKey ?: ""
            opts.systemPrompt = entity.systemPrompt ?: ""
            opts.stop = entity.stop ?: ""
            opts.contextMessageCount = entity.contextMessageCount
            opts.googleThinkingBudget = entity.googleThinkingBudget
            opts.temperature = entity.temperature
            opts.topP = entity.topP
            opts.maxTokens = entity.maxTokens
            opts.frequencyPenalty = entity.frequencyPenalty
            opts.presencePenalty = entity.presencePenalty
            opts.topK = entity.topK
            opts.streamOutput = true
            opts.autoChapterPlan = entity.autoChapterPlan
            opts.thinking = entity.thinking
            opts.autoChatEnabled = entity.autoChatEnabled
            opts.proactiveCountToday = entity.proactiveCountToday
            opts.proactiveResetDate = entity.proactiveResetDate
            opts.proactiveDailyBudget = entity.proactiveDailyBudget
            opts.inkosEnabled = entity.inkosEnabled
            opts.inkosBookId = entity.inkosBookId
            opts.inkosSubtype = entity.inkosSubtype
            opts.inkosBookRulesYaml = entity.inkosBookRulesYaml
            opts.inkosTargetChapters = entity.inkosTargetChapters
            opts.inkosChapterWordCount = entity.inkosChapterWordCount
            return opts
        }
    }
}
